lib/kafkat/command/reassign.rb

```diff
@@ -16,33 +17,139 @@ def run
           opt :brokers, "replica set (broker IDs)", type: :string
           opt :replicas, "number of replicas (count)", type: :integer
+          opt :strategy, "assignment strategy used", type: :string
         end
 
-        broker_ids = opts[:brokers] && opts[:brokers].split(',').map(&:to_i)
+        destination_broker_ids = opts[:brokers] && opts[:brokers].split(',').map(&:to_i)
         replica_count = opts[:replicas]
+        strategy = opts[:strategy]
+        strategy ||= "smart"
 
-        broker_ids ||= zookeeper.get_brokers.values.map(&:id)
+        destination_broker_ids ||= zookeeper.get_brokers.values.map(&:id)
 
-        all_brokers_id = all_brokers.values.map(&:id)
-        broker_ids.each do |id|
-          if !all_brokers_id.include?(id)
+        all_broker_ids = all_brokers.values.map(&:id)
+        destination_broker_ids.each do |id|
+          if !all_broker_ids.include? id
             print "ERROR: Broker #{id} is not currently active.\n"
             exit 1
           end
         end
 
-        # *** This logic is duplicated from Kakfa 0.8.1.1 ***
+        if strategy == "smart"
+          assignments = generate_assignment_smart(topics, destination_broker_ids,
+                                                  replica_count, all_broker_ids)
+        elsif strategy == "load-balanced"
+          assignments = generate_assignment_load_balanced(topics, destination_broker_ids,
+                                                          replica_count)
+        else
+          print "ERROR: Unrecognized strategy \"#{strategy}\".\n"
+          exit 1
+        end
+
+        prompt_and_execute_assignments(assignments)
+      end
 
+      # This is the default assignment generation strategy.
+      # There are two requirements:
+      #   1. replicas should be evenly distributed among brokers;
+      #   2. replicas moving around should be avoided as possible.
+      # We also need to address that user can sepcify both number of brokers and
+      # replication factors.
+      #
+      # To address the requirements efficiently, two passes are used to determine assignemnt.
+      # In the first pass, existing assignments where partitions with too many replicas (probably
+      # due to decrese of replication factor) and brokers with too many replicas (probably due to
+      # decrease of replication factor or increase of number of brokers) are removed.
+      # Brokers not in destination brokers are also removed from assignment in the first pass.
+      #
+      # In the second pass, partitions are assigned with available brokers if neither the partition
+      # nor the brokers are full. Assignemnts where brokers are full are swaped with availble
+      # brokers.
+      #
+      # To facilitate the two passes, two hash tables are maintained to keep track of quota
+      # on each broker and assigned partitions on each broker.
+      def generate_assignment_smart(topics, destination_broker_ids, replica_count, all_broker_ids)
```

Comment:
This function is too big. How things does it do?

Also it would nice to have a `AssignmentStrategy` class that implements a Strategy pattern (!). 
It will clearly break out the code, and make unit tests easier.